File: Litvinov group/semester 1/09.HW/hashTable.cpp
#include "hashTable.h"

#include <iostream>
#include <cmath>

using namespace std;

const int lengthOfTable = 100;
const int primeNumber = 13;

struct HashTable {
    List* tableOfLists[lengthOfTable];
};

HashTable* createHashTable() {
    HashTable* newTable = new HashTable{};
    for (int i = 0; i < lengthOfTable; i ++) {
        newTable->tableOfLists[i] = createList();
    }
    return newTable;
}

int countNumber(const string& str) {
    long long answer = 0;
    for (int i = 0; i < str.size(); ++i) {
        answer += str[i] * pow(primeNumber, i);
    }
    return answer % lengthOfTable;
}

void addString(HashTable* table, const string& str) {
    int number = countNumber(str);
    addElement(table->tableOfLists[number], str);
}

void print(HashTable* table) {
    cout << "All words mentioned in the text(word, amount): " << endl;
    for (int i = 0; i < lengthOfTable; ++i) {
        if (table->tableOfLists[i] != nullptr) {
            printList(table->tableOfLists[i]);
        }
    }
}

void calculateLength(HashTable* table) {
    int maximum = 0;
    int average = 0;
    int counter = 0;
    for (int i = 0; i < lengthOfTable; i ++) {
        int length = lengthOfList(table->tableOfLists[i]);
        if (length > maximum) {
            maximum = length;
        }
        average += length;
        if (length != 0) {
            ++counter;
        }
    }
    cout << "Maximal length of the list is: " << maximum << endl;
    cout << "Average length of the full list is: " << average / counter << endl;
    cout << "Average length of the list is: " << average / lengthOfTable << endl;
}

void deleteHashTable(HashTable* table) {
    for (int i = 0; i < lengthOfTable; ++i) {
        deleteList(table->tableOfLists[i]);
    }
    delete table;
}

File: Litvinov group/semester 2/2.HW/2.3/2.3/HashTable.cs
using static System.Math;

namespace HashTable
{
    public class HashTable
    {
        public void AddElement(string value)
        {
            int number = Abs(value.GetHashCode()) % length;
            table[number].Add(value, 1);
        }

        public bool Check(string value)
        {
            int number = Abs(value.GetHashCode()) % length;
            return table[number].CheckPresence(value);
        }

        public void Delete(string value)
        {
            int number = Abs(value.GetHashCode()) % length;
            table[number].DeleteElement(value);
        }

        public HashTable()
        {
            table = new List[length];
            for (int i = 0; i < length; ++i)
            {
                table[i] = new List();
            }
        }

        private const int length = 100;
        private List[] table;
        private const int primeNumber = 13;
    }
}
